microcloud/api/services.go

```diff
@@ -29,8 +29,9 @@ const MicroCephDir = "/var/snap/microceph/common/state"
 // Proxy returns a proxy endpoint with the given handler and access applied to all REST methods.
 func Proxy(name, path string, handler func(*state.State, *http.Request) response.Response) rest.Endpoint {
```

Comment:
Hmm, does this mean that any system with `microcloud` installed will just expose the entirety of the LXD and MicroCeph REST API to anyone who can guess its IP address?

That doesn't seem ideal. Even though the same is technically still possible by assembling a working cluster through MicroCloud and then accessing those APIs.

I'd prefer we limit the pre-init API to just the bare minimum we need to send join information.
I think we should also encode a server-specific secret in the mDNS message that needs to be transmitted when hitting the pre-init API. That way only someone on the same subnet can form use the API, anyone just randomly scanning the internet will find the listener but won't be able to use it without the secret.

When we add support for manually adding a server by address, we'll ask for both the address and that secret token to be given by the user.

Reply:
For the manual adding case, I suppose we would need a command that generates some encoded string including the address and the certificate fingerprint on uninitialized systems, that could then be specified to the existing cluster to manually add it.

Reply:
> we already have the server certificate fingerprint in the mDNS broadcast which should precede all proxied connections, so we could just add that to the request context and then on the other side we can check that it matches before proxying the request to its service. Then once the service is clustered, we can switch to checking against the cluster cert instead?

That won't work from a security standpoint as anyone can grab the server certificate by simply initializing a TLS connection.
What we need is something that only someone listening for the mDNS would be able to get.

Reply:
> For the manual adding case, I suppose we would need a command that generates some encoded string including the address and the certificate fingerprint on uninitialized systems, that could then be specified to the existing cluster to manually add it.
> 
> EDIT: So basically a string encoded version of `ServerInfo` introduced here.
> 
> ```go
> // ServerInfo is information about the server that is broadcast over mDNS.
>  type ServerInfo struct {
>  	Version     string
>  	Name        string
>  	Address     string
>  	Services    []types.ServiceType
>  	Fingerprint string
>  }
> ```

Right, for the manual add, we can just use a mostly standard token type format, includes name, addresses and the secret.

Reply:
I've now added an `AuthSecret` string that is created on each start of an uninitialized system and included in the broadcast. I didn't bother to store the secrets anywhere for now as MicroCloud doesn't currently do much after initialization. I'm thinking if we do want to do something there in the future, we would allow cluster members (recorded in the trust store) through without checking the AuthSecret, but currently it's always checked. 

The secret is stored in the request header at `X-MicroCloud-Auth` via the client proxies before the request is sent through.